suite/tests/api/ir_x86.c

```diff
@@ -787,6 +787,98 @@ test_instr_encode_and_decode(void *dc, instr_t *instr, uint len_expect,
+        return false;
+
+    /* Check that register destination operands are the same.
+     */
+    uint original_num_dsts = (uint)instr_num_dsts(instr_orig);
+    uint num_reg_dsts = 0;
+    for (uint i = 0; i < original_num_dsts; ++i) {
+        opnd_t dst_opnd = instr_get_dst(instr_orig, i);
+        if (opnd_is_reg(dst_opnd)) {
+            // TODO i#6662: we need to compare the virtual register stored in instr_copy
+            // against the converted DR_REG_ (to virtual register) of instr_orig.
+            // Currently we don't have virtual registers, so we can compare the two
+            // DR_REG_ of instr_orig and instr_copy directly.
+            if (!opnd_same(dst_opnd, instr_get_dst(instr_copy, num_reg_dsts)))
+                return false;
+            ++num_reg_dsts;
+        }
+    }
+
+    /* Check that we have the same number of destination operands that are registers.
+     */
+    if (num_reg_dsts != instr_num_dsts(instr_copy))
+        return false;
+
+    /* Check that register source operands are the same.
+     */
+    uint original_num_srcs = (uint)instr_num_srcs(instr_orig);
+    uint num_reg_srcs = 0;
+    for (uint i = 0; i < original_num_srcs; ++i) {
+        opnd_t src_opnd = instr_get_src(instr_orig, i);
+        if (opnd_is_reg(src_opnd)) {
+            // TODO i#6662: we need to compare the virtual register stored in instr_copy
+            // against the converted DR_REG_ (to virtual register) of instr_orig.
+            // Currently we don't have virtual registers, so we can compare the two
+            // DR_REG_ of instr_orig and instr_copy directly.
+            if (!opnd_same(src_opnd, instr_get_src(instr_copy, num_reg_srcs)))
+                return false;
+            ++num_reg_srcs;
+        }
+    }
+
+    /* Check that we have the same number of source operands that are registers.
+     */
+    if (num_reg_srcs != instr_num_srcs(instr_copy))
+        return false;
+
+    return true;
+}
+
+static void
+test_instr_encode_decode_synthetic(void *dc, instr_t *instr, byte *bytes)
+{
+    instr_t *instr_copy = NULL;
+    instr_copy = instr_create(dc);
+    ASSERT(instr_copy != NULL);
+    bool is_isa_mode_set = false;
+    is_isa_mode_set = instr_set_isa_mode(instr_copy, DR_ISA_SYNTHETIC);
+    ASSERT(is_isa_mode_set);
+    is_isa_mode_set = instr_set_isa_mode(instr, DR_ISA_SYNTHETIC);
+    ASSERT(is_isa_mode_set);
+    byte *next_pc = NULL;
+    next_pc = instr_encode(dc, instr, bytes);
+    ASSERT(next_pc != NULL);
+    next_pc = decode(dc, bytes, instr_copy);
+    ASSERT(next_pc != NULL);
+    ASSERT(instr_same_synthetic(instr, instr_copy));
+    instr_destroy(dc, instr);
+    instr_destroy(dc, instr_copy);
+}
+
+static void
+test_instr_create_encode_decode_synthetic(void *dc)
+{
+    instr_t *instr;
+    opnd_t abs_addr = opnd_create_abs_addr((void *)0xdeadbeefdeadbeef, OPSZ_8);
+
+    instr = INSTR_CREATE_mov_ld(dc, opnd_create_reg(DR_REG_RAX), abs_addr);
+    byte bytes[16];
```

Comment:
Move the bytes var into the helper, and have the helper assert that the decode length is < this array size.

Reply:
In the "helper" meaning from `test_instr_create_encode_decode_synthetic()` to `test_instr_encode_decode_synthetic()`?
(There are some functions named "..._helper" but I don't think they are related to what we want to do.)

Reply:
IIRC there is a call right below this line 878 and this array is passed there but the size is not passed, and the array is not used in this outer function: so logically the array should instead live in the callee who knows the size and can check for overflow.

Reply:
Yes, I changed it as requested; although I want to add more instructions to test, so I'll probably put the array that contains the raw bytes encoded instruction back in the caller, so I can reuse it for all instructions. I'll pass the size to the callee/helper so it can check for overflow.

Reply:
Not sure I understand the motivation for "reusing" the buffer: it's an uninitialized stack buffer so it has zero cost and is being reused as the stack rolls back and forth for the calls.  I would think having it local to the use for clearer code would be a positive with no downsides?

Reply:
Ah yes, it is a stack buffer, pretend I said nothing :) buffer stays in the callee.